### PR TITLE
Issue #304 Grid.onPaint is ineffient on large grids

### PR DESCRIPTION
**Pull request by laeubi**
I have don some small adjustments that for my case increases the speed dramatically at least after the final load. I'd like to provide these here as a draft to get some feedback and/or discussions.

- store index on cell to prevent linear list lookups
- don't paint headers that are out of view
- break if width is larger than view

### CODE REVIEW

File: widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java [hunk 1]
```diff
@@ -7776,9 +7784,11 @@ protected void setScrollValuesObsolete() {
 	int newColumn(final GridColumn column, final int index) {
 
 		if (index == -1) {
+			column.index = columns.size();
 			columns.add(column);
 			displayOrderedColumns.add(column);
 		} else {
+			column.index = index;
```
Comment (tomsontom) [inline widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java:7792]: you need to sync all indices after that one

File: widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java [hunk 2]
```diff
@@ -7776,9 +7784,11 @@ protected void setScrollValuesObsolete() {
 	int newColumn(final GridColumn column, final int index) {
```
Comment (tomsontom) [inline widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java:7784]: beside handling the `newColumn` you also have to handle the `removeColumn` and fix all indices

**Review by tomsontom — CHANGES_REQUESTED** [review]

**Comment by laeubi** [comment]
FYI @wimjongman as I can't add reviewer to a PR...

File: widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java [hunk 3]
```diff
@@ -5447,24 +5450,29 @@ private void onPaint(final PaintEvent e) {
...
 					if (column.isVisible()) {
-						emptyCellRenderer.setBounds(x, y, column.getWidth(), getItemHeight());
-						emptyCellRenderer.setColumn(indexOf(column));
-						emptyCellRenderer.paint(e.gc, this);
-
-						x += column.getWidth();
+						int width = column.width;
+						if(x + width >= 0) {
+							emptyCellRenderer.setBounds(x, y, width, getItemHeight());
+							emptyCellRenderer.setColumn(indexOf(column));
```
Comment (tomsontom) [inline widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java]: you could use the cached value where to save the method call and the checks

File: widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java [hunk 4]
```diff
@@ -5447,24 +5450,29 @@ private void onPaint(final PaintEvent e) {
...
 
 					if (column.isVisible()) {
-						emptyCellRenderer.setBounds(x, y, column.getWidth(), getItemHeight());
-						emptyCellRenderer.setColumn(indexOf(column));
-						emptyCellRenderer.paint(e.gc, this);
-
-						x += column.getWidth();
+						int width = column.width;
+						if(x + width >= 0) {
+							emptyCellRenderer.setBounds(x, y, width, getItemHeight());
```
Comment (tomsontom) [inline widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java]: you could move the getItemHeight() out of the loop or access it directly to by pass the check

File: widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java [hunk 5]
```diff
@@ -5447,24 +5450,29 @@ private void onPaint(final PaintEvent e) {
 					x += rowHeaderWidth;
 				}
 
-				emptyCellRenderer.setBounds(x, y, getClientArea().width - x, getItemHeight());
+				emptyCellRenderer.setBounds(x, y, clientArea.width - x, getItemHeight());
```
Comment (tomsontom) [inline widgets/grid/org.eclipse.nebula.widgets.grid/src/org/eclipse/nebula/widgets/grid/Grid.java]: you can directly access the getItemHeight() or save it and reuse it below

**Review by tomsontom — CHANGES_REQUESTED** [review]

**Comment by wimjongman** [comment]
Moving to that Tycho version caused an issue. I think the dependency has to be added explicitly to the manifest of grid.css

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.2.0:compile (default-compile) on project org.eclipse.nebula.widgets.grid.css: Compilation failure: Compilation failure: 
[ERROR] /home/jenkins/agent/workspace/nebula.stable.github/widgets/grid/org.eclipse.nebula.widgets.grid.css/src/org/eclipse/nebula/widgets/grid/css/GridElementProvider.java: 
[ERROR] 	/*******************************************************************************
[ERROR] 	^
[ERROR] The type org.w3c.dom.css.CSSValue cannot be resolved. It is indirectly referenced from required .class files
```

**Comment by wimjongman** [comment]
@laeubi, Please take a look at the failing build.

**Review by tomsontom — APPROVED** [review]

**Comment by laeubi** [comment]
Could this be merged or is anything left?

**Comment by tomsontom** [comment]
as I'm not directly involved anymore I let @wimjongman merge it - i only looked through the code and can't comment on build, versioning stuff, ...

**Pull request merged**